src/lib/parser.c

```diff
@@ -1664,6 +1674,191 @@ RdbStatus elementSetLP(RdbParser *p) {
+
+    int ret = ziplistValidateIntegrity(ziplistBulk->ref, ziplistBulk->len, p->deepIntegCheck, NULL, NULL);
+
+    if (unlikely(!ret)) {
+        RDB_reportError(p, RDB_ERR_ZSET_ZL_INTEG_CHECK, "zsetZiplistItem(): Ziplist integrity check failed");
+        return RDB_STATUS_ERROR;
+    }
+
+    if (p->elmCtx.key.handleByLevel == RDB_LEVEL_STRUCT) {
+        registerAppBulkForNextCb(p, ziplistBulk);
+        CALL_HANDLERS_CB(p, NOP, RDB_LEVEL_STRUCT, rdbStruct.handleZsetZL, ziplistBulk->ref);
+        return RDB_STATUS_OK;
+    }
+
+    unsigned char *offsetZL = ziplistIndex(ziplistBulk->ref, 0);
+    while (offsetZL != NULL) {
+        unsigned char *item1, *item2;
+        unsigned int item1Len, item2Len;
+        long long item1Val, item2Val;
+        EmbeddedBulk embBulk;
+
+        ziplistGet(offsetZL, &item1, &item1Len, &item1Val);
+        offsetZL = ziplistNext(ziplistBulk->ref, offsetZL);
+
+        ziplistGet(offsetZL, &item2, &item2Len, &item2Val);
+        offsetZL = ziplistNext(ziplistBulk->ref, offsetZL);
+
+        double score;
+        if (item2) {
+            if (!string2d((const char*) item2, item2Len, &score)) {
+                RDB_reportError(p, RDB_ERR_ZSET_ZL_INTEG_CHECK,
+                                "elementZsetZL(): failed to parse string to double: %.*s",
+                                item2Len, item2);
+
+                return RDB_STATUS_ERROR;
+            }
+        } else {
+            score = (double) item2Val;
+        }
+
+        if (!allocEmbeddedBulk(p, item1, item1Len, item1Val, &embBulk))
+            return RDB_STATUS_ERROR;
+
+        registerAppBulkForNextCb(p, embBulk.binfo);
+        CALL_HANDLERS_CB(p,
+                         restoreEmbeddedBulk(&embBulk);, /*finalize*/
+                         RDB_LEVEL_DATA,
+                         rdbData.handleZsetMember,
+                         embBulk.binfo->ref,
+                         score);
+    }
+    return RDB_STATUS_OK;
+}
+
+RdbStatus elementZsetZL(RdbParser *p) {
+    BulkInfo *ziplistBulk;
+
+    IF_NOT_OK_RETURN(rdbLoadString(p, RQ_ALLOC_APP_BULK, NULL, &ziplistBulk));
+
+    /*** ENTER SAFE STATE ***/
+
+    if (RDB_STATUS_ERROR == zsetZiplistItem(p, ziplistBulk))
+        return RDB_STATUS_ERROR;
+
+    return nextParsingElement(p, PE_END_KEY);
+}
+
+RdbStatus elementZsetLP(RdbParser *p) {
+    unsigned char *iterator, *item1, *item2;
+    unsigned int item1Len, item2Len;
+    long long item1Val, item2Val;
+    BulkInfo *listpackBulk;
+
+    IF_NOT_OK_RETURN(rdbLoadString(p, RQ_ALLOC_APP_BULK, NULL, &listpackBulk));
+
+    /*** ENTER SAFE STATE ***/
+
+    /* Doesn't check for duplication */
+    if (!lpValidateIntegrity(listpackBulk->ref, listpackBulk->len, p->deepIntegCheck, NULL, 0)) {
+        RDB_reportError(p, RDB_ERR_ZSET_LP_INTEG_CHECK, "elementZsetLP(): LISTPACK integrity check failed");
+        return RDB_STATUS_ERROR;
+    }
+
+    /* TODO: handle empty listpack */
```

Comment:
I tested that redis 7.x doesn't save to rdb zset with 0 members. Please double check
It is the same for old version and if so, please delete this comment. Otherwise, let's
think how to change the code to support it.

Reply:
Deleted. I already tried it and this is just a copy-paste leftover.